lib/std/math.zig

```diff
@@ -97,80 +93,163 @@ pub fn approxEqAbs(comptime T: type, x: T, y: T, tolerance: T) bool {
 /// Note that for comparisons of small numbers around zero this function won't
 /// give meaningful results, use `approxEqAbs` instead.
 ///
-/// NaN values are never considered equal to any value.
+/// NaN values are never equal to any value, including themselves.
 pub fn approxEqRel(comptime T: type, x: T, y: T, tolerance: T) bool {
-    assert(@typeInfo(T) == .Float or @typeInfo(T) == .ComptimeFloat);
-    assert(tolerance > 0);
-
-    // Fast path for equal values (and signed zeros and infinites).
-    if (x == y)
-        return true;
-
-    if (isNan(x) or isNan(y))
-        return false;
+    switch (@typeInfo(T)) {
+        .Float, .ComptimeFloat => {},
+        else => @compileError("T must be float or comptime float."),
+    }
+    assert(tolerance >= 0);
+    if (x == y) return true;
+    if (isNan(x) or isNan(y)) return false;
+    return @abs(x - y) <= tolerance * @abs(0.5 * x + 0.5 * y);
+}
 
-    return @abs(x - y) <= @max(@abs(x), @abs(y)) * tolerance;
+/// Performs an approximate comparison of two floating point values `x` and `y`.
+/// Returns true if the number of representable intervals (gaps) between them is
+/// less or equal than the specified count.
+///
+/// For example, two numbers with three representable values (points)
+/// between them are four intervals (gaps) apart from each other, i.e.
+///
+/// ```
+/// ... a - * - * - * - b ...
+///     | 1 | 2 | 3 | 4 |
+/// ```
+///
+/// The `ulp` parameter is the maximum permissible number of representable
+/// floating point intervals between the two numbers being compared.
+///
+/// Numbers on opposite sides of zero are compared on a scale with the
+/// denormalized numbers comprising a linear crossover region around zero,
+/// where positive and negative zeros are mapped to the same point and do not
+/// comprise an extraneous interval, i.e.
+///
+/// ```
+/// | -NaN | -inf | -nor | -den | 0 | +den | +nor | +inf | +NaN |
+/// ```
+///
+/// NaN values are never equal to any value, including themselves.
+pub fn approxEqUlp(comptime T: type, x: T, y: T, ulp: comptime_int) bool {
```

Comment:
`ulp` appears to be comptime for no reason

doc comments don't mention what "ulp" stands for anywhere.